doc/audioplayer.de.man
.\" Process this file with
.\" groff -man -Tutf8 function_plotter.de.man
.\"
.
.TH "audioplayer.lua" 1 2012-07-25 AKFAvatar
.nh
.
.SH NAME
audioplayer.lua \- Audio-Abspieler f\[:u]r AKFAvatar
.
.SH SYNTAX
.BI "audioplayer.lua [" "Audio-Datei(en)" ]
.
.SH BESCHREIBUNG
Audio-Abspieler f\[:u]r AKFAvatar.
.PP
.TP
Unterst\[:u]tzte Audio-Formate:
Ogg Vorbis, Wave, AU
.TP
Unterst\[:u]tzte Playlisten-Formate:
M3U, PLS
.PP
Dieses Programm kann entweder interaktiv verwendet werden,
oder Dateien k\[:o]nnen als Parameter \[:u]bergeben werden.
.PP
Wenn man es von einem anderen Lua-Programm mit
.B dofile()
aufrufen will,
kann man die globale Variable
.I AUDIO
verwenden. Diese sollte entweder eine Zeichenkette,
oder eine Tabelle mit Zeichenketten enthalten.
.PP
Ogg Vorbis wird nur unterst\[:u]tzt, wenn das Modul \[Bq]akfavatar-vorbis\[lq]
verf\[:u]gbar ist.
Es k\[:o]nnen keine komplexen Dateien mit mehreren Streams abgespielt werden,
nur \[Bq]Vorbis I simple profile\[lq].
.PP
Wenn man \[Bq]play all...\[lq] ausw\[:a]hlt, oder mehrere Dateien als Parameter
\[:u]bergibt, dann werden Playlisten ignoriert.
Playlisten k\[:o]nnen auch nicht auf andere Playlisten verweisen.
.PP
Netzwerk-Adressen werden unterst\[:u]tzt, wenn das Programm \[Bq]curl\[lq] im
.B PATH
installiert ist.
Aber das funktioniert nicht auf allen Systemen.
.PP
Wenn in dem Verzeichnis mit den Dateien eine Bild-Datei namens \[Bq]cover\[lq]
liegt, wird dieses Bild bei der Wiedergabe angezeigt.
Wenn man eine Playliste mit relativen Pfaden oder URLs verwendet,
dann muss das Bild im selben Verzeichnis wie die Playliste liegen.
Das Bild wird auf keinen Fall vom Netz geladen.
.PP


NEWS
* AKFAvatar-testing

 - optimized printing of text (line buffered)
 - audio-data with same specs as former one is played with less latency
 - improved support for 24 and 32 bit audio
 - new file Bücher.lua for reading German books from my server (needs curl)

  C-API changes:
    - new macro: AVT_KEY_F

* AKFAvatar 0.24.3

  - audio: big audio files are read directly from file while playing
    (if possible via mmap)
  - audio: AU and WAV data can be read from non-seekable streams
    (loaded into memory)
  - fix: Pascal sound support fixed
  - the control characters SO/SI switch underlined mode on/off;
    the control characters DC1/DC2 switch bold mode on/off
  - configure: new option --with-sdl2
  - SDL2: support for SDL2 with automatic resizing of the graphics
  - SDL2: support for pasting text from clipboard (Alt+V, middle mouse button)

  Lua-API changes:
    - new functions: avt.frequency
    - module "akfavatar-term": fix APC interface

* AKFAvatar 0.24.2

  - lua/clock.lua: resizing works (with some delay)
  - lua/four_in_a_row.lua: mouse support / resizing
  - lua/four_in_a_row.lua: play against computer

  Lua-API changes:
    - Module "akfavatar-graphic":
      - new functions: graphic.set_pointer_buttons_key,
        graphic.set_pointer_motion_key, graphic.get_pointer_position,
        graphic.set_resize_key

* AKFAvatar 0.24.1

  - fix: lua/interactive_lua.lua: fix handling of incomplete lines
  - Lua module "akfavatar-graphic" included in executable
  - text cursor modified
  - lua-akfavatar: GNU .info files can be shown as plain text (no hyperlinks)
  - lua/hexviewer.lua: show more characters on the right side
  - lua/multiplication.lua: clear board (no scrolling)
  - configure accepts the option --with-linuxfb
    which then builds for the Linux framebuffer without the SDL
    (no mouse and no audio support yet)

  C-API changes:
    - new functions: avt_detect_utf8

  Lua-API changes:
    - new functions: avt.toutf8, avt.detect_utf8, avt.utf8codepoints

* AKFAvatar 0.24.0

  - fix: terminal doesn't use so much CPU time anymore
  - fix: window resizing for the pager works again
  - fix: switching fullscreen mode while avatar moves in or out
  - fix: avt_credits sometimes got the end wrong
  - fix: audioplayer.lua accepts PLS playlists again
  - fix: Pascal binding works again, extended
  - paging through menus is animated
  - lua/audioplayer.lua: improved playlist support
  - lua/multiplication.lua: more authentic

  C-API changes:
    - avt_char is unsigned int now, avt_choice adapted
      should be compatible with char32_t in C11
    - functions with *_mb* replaced with functions *_char*
    - new functions: avt_darker, avt_brighter, avt_brightness
    - avtadons.h also uses the prefix avt_ for everything

  Lua-API changes:
    - The default encoding is now ASCII
      to encourage always using avt.encoding()
    - avt.set_directory or avt.chdir return true on success
      or nil and an error message on error

  Pascal-API changes:
    - variable CheckEsc removed
    - new procedures/functions: Menu, ReadKeyCode, reserve_single_keys


* AKFAvatar 0.23.1

  - fixed vulnerabilities
  - avoid recursive calls of the bell
  - split source code for better maintainablitity
  - name sign with 3d effect

  C-API changes:
    - avt_set_title: no automatic charset conversion
    - avt_put_char: support for UTF-16 surrogate characters
    - new functions: avt_wide_copyright, avt_wide_license, avt_wide_version,
      avt_menu, avt_palette_size
    - avtaddons:
      - avta_file_selection and avta_filter have a new parameter "data",
        which is passed along

  Lua-API changes:
    - Module "akfavatar-graphic"
      - new methods: gr:border3d

* AKFAvatar 0.23.0

  The code was made less dependend on SDL. There is a new experimantal
  backend for a Linux framebuffer device, but without mouse support or audio
  (of course you can call an external audio-player in your scripts).

  - bugfix: fix for XPM images with more than 256 colors
  - bugfix: lua-akfavatar also resets the charset
  - bugfix/change: lua-akfavatar option -l sets a global variable
  - avt_update() is much faster
  - support for combining characters (Unicode)
  - The pager supports numbers (number-pad when NumLock is on)
  - lua/interactive_lua.lua supports scrolling back in a history list
  - regression: the navigation bar cannot handle the pause key anymore
  - key codes have been changed not to conflict with the "corporate zone"

  C-API changes:
    - all deprecated functions removed
    - also removed: avt_register_keyhandler, avt_register_mousehandler
    - avt_update() is much faster
    - avt_get_key replaces avt_key
    - modified: avt_show_raw_image, avt_put_raw_image_file,
      avt_put_raw_image_stream, avt_put_raw_image_data, avt_put_raw_image_xpm
      don't have the bpp parameter anymore, only 4 Bytes per pixel accepted
    - new functions: avt_key_pressed, avt_push_key, avt_clear_keys,
      avt_set_audio_end_key, avt_combining, avt_bell_function
    - new experimental functions: avt_set_pointer_buttons_key,
      avt_set_pointer_motion_key, avt_get_pointer_position,
      avt_char avt_input, avt_input_mb
    - new macro: avt_elapsed

  Lua-API changes:
    - the function avt.ask was extended (yet undocumented)
    - new functions: avt.combining, avt.key_pressed, avt.clear_keys,
      avt.push_key, avt.set_audio_end_key, avt.alert, avt.getcwd
    - new table: avt.key with key codes


* AKFAvatar 0.22.1

  There are a lot of changes in the APIs. Also the ABI was changed.

  - Window is resizable again (content gets centered)
  - The image can be a separated banner as header or footer
  - Uses Lua-5.2 now, no longer compatible with 5.1
  - Lua modules don't define globals anymore (scripts have to be adapted!)
  - Data files are put in a separate directory and Lua has a search function
    for it (scripts have to be adapted!)
  - the function for translations is integrated in Lua-AKFAvatar

lua/audioplayer.lua
#!/usr/bin/env lua-akfavatar

--[[---------------------------------------------------------------------
Audio player for AKFvatar
Copyright (c) 2011,2012,2013 Andreas K. Foerster <akf@akfoerster.de>
License: GPL version 3 or later

Supported audio formats: Ogg Vorbis, Wave, AU
Supported decoders: flac, opusdec, mpg123, (mpg321, madplay)
Supported playlist formats: M3U, PLS
--]]--------------------------------------------------------------------

local avt = require "lua-akfavatar"
local vorbis = avt.optional "akfavatar-vorbis"
local default_cover = assert(avt.search("audio1.xpm"))

-- use $input for input file name, $output for AU or WAVE output files
-- comment out if you don't even want those files to show up
local decoder_flac = 'flac --decode --silent --force --output-name="$output" "$input"'
local decoder_opus = 'opusdec --quiet --force-wav "$input" "$output"'
local decoder_mpeg = 'mpg123 --quiet --au "$output" "$input"'
--local decoder_mpeg = 'mpg321 --quiet --au "$output" "$input"'
--local decoder_mpeg = 'madplay -q -o "wave:$output" "$input"'
-- note: madplay supports only 8 bit for the snd format, so avoid that

-- downloader application
-- url is appended, data should be dumped to stdout
local downloader = "curl --silent --location"

avt.translations = {
  ["Audio-Player"] = {
    de="Audio-Abspieler"},

  ["play all..."] = {
    de="alle abspielen..."},
}

local L = avt.translate

avt.title(L"Audio-Player")
avt.start()
avt.start_audio()
avt.set_balloon_color("tan")
avt.encoding("UTF-8")

local function tempname()
  local tmp = os.tmpname()

  -- for Windows
  if string.find(tmp, "^\\") then
    tmp = os.getenv("TMP") .. tmp
  end

  -- print ("tmp:", tmp)

  return tmp
end

-- open URL with the tool "curl"
-- returns file handle
local function open_url(url)
  -- Windows needs the "rb" mode - most system don't accept it
  return assert(io.popen(downloader .. " " .. url, "rb")
            or io.popen(downloader .. " " .. url, "r"))
end

local function get_url(url) -- might not work on Windows
  local f = open_url(url)
  local data = f:read("*all")
  f:close()
  return data
end

local function show_cover(dir)
  dir = dir or ""
  if not avt.show_image_file(dir.."cover") then
    avt.show_image_file(default_cover)
  end
end

-- replace file-URL with normal filename and remove \r
local function handle_list_entry(s)
  s = string.gsub(s, "\r", "")
  s = string.gsub(s, "^file://localhost/(%a:.*)$", "%1") -- Windows
  s = string.gsub(s, "^file:///(%a:.*)$", "%1") -- Windows
  s = string.gsub(s, "^file://localhost/", "/")
  s = string.gsub(s, "^file:///", "/")
  return s
end

local function load_decoder(decoder, url)
  local tmp = tempname()

  local decode = string.gsub(decoder, "%$(%w+)",
                   {output=tmp, input=handle_list_entry(url)})

  local audio = nil
  if os.execute(decode) then
    audio = avt.load_audio_file(tmp, "play")
  end

  -- remove a file while it is opened works on most systems
  -- it is just really removed after closing
  os.remove(tmp)

  return audio
end

local function load_audio(url)
  local audio = nil
  local u = string.lower(url)

  if string.find(u, "^https?://") or string.find(u, "^s?ftps?://") then
    audio = avt.load_audio(get_url(url), "play")
  elseif string.find(u, "%.flac?$") then
    audio = load_decoder(decoder_flac, url)
  elseif string.find(u, "%.opus$") then
    audio = load_decoder(decoder_opus, url)
  elseif string.find(u, "%.mp[123]$") then
    audio = load_decoder(decoder_mpeg, url)
  else
    audio = avt.load_audio_file(handle_list_entry(url), "play")
  end

  return audio
end

local function play_single(filename) --> play a single file
  local audio, button

  show_cover(string.match(filename, "^(.-)[^\\/]+$"))

  audio = load_audio(filename)

  if not audio then
    avt.tell("cannot play audio file")
    avt.wait_button()
    return
  end

  repeat
    button = avt.navigate("ps")
    if button == "p" then
      avt.pause_audio(true)
      button = avt.navigate("rs")
      if button == "r" then avt.pause_audio(false) end
    end
  until button == "s" -- until stop is pressed or sound stopped

  avt.stop_audio()
  audio:free()
end -- play_single

-- splits a filename into the name and the directory
-- problem: doesn't work on URLs
local function basename(filename)
  local dir, name = string.match(filename, "^(.-)([^\\/]+)$")
  return name, dir
end

local function m3u(url) --> M3U Playlist
  local list = {}
  local titles = {}
  local file

  if string.find(url, "^https?://") or string.find(url, "^s?ftps?://") then
    file = open_url(url)
  else --> local file
    local name, dir = basename(url)
    avt.set_directory(dir) --> list may contain relative paths
    file = assert(io.open(name))
  end

  for line in file:lines() do
    if not string.find(line, "^%s*#") then 
      table.insert(list, handle_list_entry(line))
    else
      local title = string.match(line, "^%s*#EXTINF:%d*,%s*(.*)$")
      if title then
        table.insert(titles, title)
      end
    end
  end

  file:close()

  if titles == {} then
    titles = nil
  end

  return list, titles
end

local function pls(url) --> PLS ShoutCast Playlist
  local list = {}
  local titles = {}
  local file

  if string.find(url, "^https?://") or string.find(url, "^s?ftps?://") then
    file = open_url(url)
  else --> local file
    local name, dir = basename(url)
    avt.set_directory(dir) --> list may contain relative paths
    file = assert(io.open(name))
  end

  if not string.match(file:read("*line"), "^%[playlist%]%s*$") then
    file:close()
    return nil
    end

  for line in file:lines() do
    local number, path = string.match(line, "^%s*File(%d+)=(.*)$")
    if number and path then
      number = tonumber(number)
      path = handle_list_entry(path)
      list[number] = path
      if not titles[number] then
        titles[number] = path
      end
    else --> not a File entry
      local title
      number, title = string.match(line, "^%s*Title(%d+)=(.*)$")
      if number and title then
        titles[tonumber(number)] = title
      end
    end
  end

  file:close()

  if titles == {} then
    titles = nil
  end

  return list, titles
end

local function play_list(list, titles) --> plays a list of files (but no playlists)
  local filename, audio, button
  local number = 1

  -- no list?
  if not list or not list[1] then return end

  -- just one entry?
  if not list[2] then return play_single(list[1]) end

  -- eventually show playlist
  ::playlist::
  if titles then
    avt.set_balloon_size (rawlen (titles), 0)
    number = avt.menu(titles)
  end

  show_cover()

  while list[number] and button ~= "s" and button ~= "x" do

    repeat -- get entry
      filename = list[number]
      if not filename then break end
      if audio then audio:free() end
      audio = load_audio(filename)
      -- remove unplayable entries
      if not audio then table.remove(list, number) end
    until audio or not list[number]

    if not audio then return end

    repeat
      if titles then
        button = avt.navigate("bfpsx")
      else
        button = avt.navigate("bfps")
      end

      if button == "p" then
        avt.pause_audio(true)
        if titles then
          button = avt.navigate("bfrsx")
        else
          button = avt.navigate("bfrs")
        end
        if button == "r" then avt.pause_audio(false) end
      end

      if button == "b" then number = number - 1
      elseif button == "f" then number = number + 1
      end
    until button ~= "r" --> repeat if play pressed

  end --> while list[number]

  avt.stop_audio()
  if audio then audio:free() end

  if titles and button ~= "x" then
    button = nil
    goto playlist
  end
end -- play_list

local function play(e) --> play file or list
  local t = type(e)

  if "table" == t then
    play_list(e)
  elseif "string" == t then
    if string.find(string.lower(e), "%.m3u$") then
      play_list(m3u(e))
    elseif string.find(string.lower(e), "%.pls$") then
      play_list(pls(e))
    else
     play_single(e)
    end
  -- do nothing for nil or other types
  end
end

-- which files to show in the file chooser
local function supported_file(n)
  n = string.lower(n)
  return (vorbis and string.find(n, "%.ogg$"))
         or (decoder_flac and string.find(n, "%.flac?$"))
         or (decoder_opus and string.find(n, "%.opus$"))
         or (decoder_mpeg and string.find(n, "%.mp[123]$"))
         or string.find(n, "%.au$")
         or string.find(n, "%.snd$")
         or string.find(n, "%.wav$")
         or string.find(n, "%.m3u$")
         or string.find(n, "%.pls$")
end

local function play_select()
  local name

  -- go home
  avt.set_directory(os.getenv("HOME") or os.getenv("USERPROFILE"))

  local files = avt.directory_entries()
  table.sort(files)

  repeat
    local audio_files = {}
    local menu = {
      {"\xE2\x97\x84", "/up"},
      {"\xE2\x80\xA2 " .. L"play all...", "/all"}
      }

    -- find supported entries
    for i, v in ipairs(files) do
      if not string.find(v, "^%.") then
        if supported_file(v) then
          table.insert(audio_files, v)
          table.insert(menu, {avt.recode(v, "SYSTEM"), v})
        elseif avt.entry_type(v) == "directory" then
          table.insert(menu, {avt.recode(v, "SYSTEM") .. "\xE2\x96\xBA",
                              v .. "/"})
        end
      end
    end

    avt.set_balloon_size(0, 0)
    avt.bold(true)
    avt.say(L"Audio-Player")
    avt.bold(false)
    avt.newline()

    name = avt.long_menu(menu)

    avt.show_avatar()

    if name == "/up" then
      avt.set_directory("..")
      files = avt.directory_entries()
      table.sort(files)
    elseif name == "/all" then
      play_list(audio_files) --> sorts out playlists
    elseif string.find(name, "/$") then -- directory
      avt.set_directory(name)
      files = avt.directory_entries()
      table.sort(files)
    else
      play(name)
    end
  until not name
end

if AUDIO then
  play(AUDIO) -- global AUDIO may be a string or an array
elseif arg[2] then
  play(arg) -- more than one file
elseif arg[1] then
  play(arg[1]) -- single file
else
  play_select()
end
